### ISSUE TITLE
meet cub error when making

### ISSUE THREAD

**Issue by Mathew0553** [limbo018/DREAMPlace#141]
Dear professor，i have received error when making:
```
/usr/include/cub/util_namespace.cuh:46:2: error: #error CUB requires a definition of CUB_NS_QUALIFIER when CUB_NS_PREFIX/POSTFIX are defined.
   46 | #error CUB requires a definition of CUB_NS_QUALIFIER when CUB_NS_PREFIX/POSTFIX are defined.
      |  ^~~~~
```
although this error can be search on the google , i tried almost each kinds of method, but still doesn't work for me,
maybe it should change somewhere in `util_namespace.cuh` ?

please kindly help! it is important for me,thanks

**Comment by limbo018** [comment]
It looks like you have a separate cub installation under ```/usr/include```. What's your CUDA version and where is it installed?

**Comment by Mathew0553** [comment]
> It looks like you have a separate cub installation under `/usr/include`
 yes,i'm just one of the student users in my laboratory，cub is installed on `/usr/include ` but not used the cub in my conda list

>What's your CUDA version and where is it installed?
CUDA version is 11.7 and it is installed on `/usr/local`
how can i fix it?
thanks for reply!

**Comment by limbo018** [comment]
CUDA 11.7 has already integrated cub. So the other cub under ```/usr/include``` conflicts with it. You might need to try 
1. Move the other cub somewhere else so that it is not included. 
2. Exclude ```/usr/include``` from your ```PATH``` environment, but this may cause other problems.

**Comment by limbo018** [comment]
Hi, are you work with the latest code? I checked the header file https://github.com/limbo018/DREAMPlace/blob/master/dreamplace/ops/utility/src/utils_cub.cuh and found `CUB_NS_QUALIFIER` is already defined.

**Comment by Mathew0553** [comment]
Thanks for your patients
okay, i was told to used DREAMPlace3.0.0，
Is there a significant difference between the two versions, and can the latest code replace version 3.0.0？
Besides, i tried DREAMPlace latest code and meet another Unsolvable torch problem when cmake ..
![1](https://github.com/limbo018/DREAMPlace/assets/113444740/edc91746-0c49-471f-adbd-1d8ff7c5b4a2)
i got this with the same environment..

if lastest code is different than previous，i would like to fix the version 3.0.0 bug. Do you have any good suggestions？
@limbo018

**Comment by limbo018** [comment]
The code on master branch is version 4.0.0 which supports timing-driven optimization (off by default). If you want to fix the problem with version 3.0.0, you can follow the contents in https://github.com/limbo018/DREAMPlace/blob/master/dreamplace/ops/utility/src/utils_cub.cuh modify where cub is included.

**Comment by limbo018** [comment]
For the error with version 4.0.0, you didn't pull the submodules. 
```
git submodules init
git submodules update
```

**Comment by Mathew0553** [comment]
> For the error with version 4.0.0, you didn't pull the submodules.
> 
> ```
> git submodules init
> git submodules update
> ```

i execute `git clone --recursive https://github.com/limbo018/DREAMPlace.git` but sill got the same  torch error , it can not found the correct torch and python.
![Uploading 1.png…]()

**Comment by limbo018** [comment]
Can you try `-DPython_EXECUTABLE`? Another issue reports that the new FindPython in cmake has changed the API to case-sensitive.

**Comment by Mathew0553** [comment]
i would like to know can i download version 3.0 through `git clone --recursive`?

**Comment by Mathew0553** [comment]
Sincerely thank you for all your help and patience. I can build the latest code now. Wish u best !@limbo018

**Comment by limbo018** [comment]
> i would like to know can i download version 3.0 through `git clone --recursive`?

You can try following command according a tutorial for git (https://devconnected.com/how-to-checkout-git-tags/)
```
git checkout tags/3.0.0
```

**Comment by Mathew0553** [comment]
> git checkout tags/3.0.0
Hi, professor Lin , I have try your advice and correctly get the 3.0.0version code and its submodules, but still got the cub error,i think the thirdparty's cub has not changed between 4.0 and 3.0. 
But never mind , currently i successfully using the 4.0 version.
Now , i have another question: 
![1](https://github.com/limbo018/DREAMPlace/assets/113444740/f1cf6863-e2bd-4076-8db1-2dcc7bcecb1f)
when i wanna check `place_io_cpp.pydb`’s content, find out if it provides standard cell position（including node_x ,node_y,node_name and so on）. i found that `make install` step created the `place_io.cpp.so` executable file but not the cpp file. When I used to build, I also generated executable file instead of cpp file. how do i understand that?
i checked <[#39](https://github.com/limbo018/DREAMPlace/issues/39)> But it doesn't seem like the same issue.
Thanks for reply.

**Comment by limbo018** [comment]
You might have misunderstanding about what `make` and `make install` do. It might be helpful to take a look at the basics of compiling and linking process. 
https://www.cs.cmu.edu/~213/lectures/15-linking.pdf

In a CMake building system (refer to CMake docs), `make` does the real compiling and linking, while `make install` only copies files to where you want to install.

**Comment by Mathew0553** [comment]
> > i would like to know can i download version 3.0 through `git clone --recursive`?
> 
> You can try following command according a tutorial for git (https://devconnected.com/how-to-checkout-git-tags/)
> 
> ```
> git checkout tags/3.0.0
> ```

dear limbo professors,
sorry for bothering you,  i would like to ask you for device :recenlty i do my experiment , i must use DREAMPlace-3.0.0,  but still meet  the  below torch problem when i cmake,  i can fix it by myself , could you please help me  fix it?
![Uploading problem.png…]()
